Cirrious/Cirrious.MvvmCross/ViewModels/MvxAsyncCommand.cs

```diff
@@ -0,0 +1,283 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Cirrious.CrossCore;
+using Cirrious.CrossCore.ExtensionMethods;
+using Cirrious.CrossCore.Platform;
+
+namespace Cirrious.MvvmCross.ViewModels
+{
+    public abstract class MvxAsyncCommandBase
+        : MvxCommandBase
+    {
+        private readonly object _syncRoot = new object();
+        private readonly bool _allowConcurrentExecutions;
+        private CancellationTokenSource _cts;
+        private int _concurrentExecutions;
+
+        protected MvxAsyncCommandBase(bool allowConcurrentExecutions = false)
+        {
+            _allowConcurrentExecutions = allowConcurrentExecutions;
+        }
+
+        public event EventHandler<MvxCommandErrorEventArgs> ErrorOccured;
+
+        public bool IsRunning
+        {
+            get { return _concurrentExecutions > 0; }
+        }
+
+        protected CancellationToken CancelToken 
+        {
+            get 
+            {
+                return _cts.Token; 
+            }
+        }
+
+        protected abstract bool DoCanExecute(object parameter);
+        protected abstract Task DoExecuteAsync(object parameter);
+
+        public void Cancel()
+        {
+            lock (_syncRoot)
+            {
+                if (_cts == null)
+                {
+                    Mvx.Trace(MvxTraceLevel.Warning, "MvxAsyncCommand : Attempt to cancel a task that is not running");
+                }
+                else
+                {
+                    _cts.Cancel();
+                }
+            }
+        }
+
+        public bool CanExecute()
+        {
+            return CanExecute(null);
+        }
+
+        public bool CanExecute(object parameter)
+        {
+            if (!_allowConcurrentExecutions && IsRunning)
+                return false;
+            else
+                return DoCanExecute(parameter);
+        }
+
+        public async void Execute(object parameter)
+        {
+            await ExecuteAsync(parameter).ConfigureAwait(false);
+        }
+
+        public async void Execute()
```

Comment:
Exceptions thrown in `async void` methods couldn't be catched. While `DoExecute` calls are wrapped in `try/catch` blocks, `DoCanExecute` are not, so any exception thrown there might crash the application. So, the question is, do we really need synchronous `Execute` methods in a class, which made to be asyncronous? If yes, I believe, it will be better to wrap `DoCanExecute` calls in a `try/catch` block too.